### ISSUE TITLE
[BUG] Problems to build unbound with nghttp2 and OpenSSL 3.3.0 / 3.2.1

### ISSUE THREAD

**Issue by iz8mbw** [NLnetLabs/unbound#1063]
I have problems to build `unbound` and this happens when using already pre built `nghttp2` library.
In details I have `nghttp2` built with `OpenSSL 3.3.0` in `/usr/local/lib` and when I try to build `unbound` linking `nghttp2` and linking the same `OpenSSL 3.3.0` I have this error during `make`:

```
./libtool --tag=CC --mode=link gcc -R/opt/ssl33/lib64 -I.  -I/opt/ssl33/include -I/usr/local/include -DSRCDIR=. -g -O2 -L/opt/ssl33/lib64 -L/usr/local/lib -version-info 9:26:1 -no-undefined -export-symbols ./libunbound/ubsyms.def -o libunbound.la context.lo libunbound.lo libworker.lo ub_event_pluggable.lo dns.lo infra.lo rrset.lo dname.lo msgencode.lo as112.lo msgparse.lo msgreply.lo packed_rrset.lo iterator.lo iter_delegpt.lo iter_donotq.lo iter_fwd.lo iter_hints.lo iter_priv.lo iter_resptype.lo iter_scrub.lo iter_utils.lo localzone.lo mesh.lo modstack.lo view.lo outbound_list.lo alloc.lo config_file.lo configlexer.lo configparser.lo fptr_wlist.lo siphash.lo edns.lo locks.lo log.lo mini_event.lo module.lo net_help.lo random.lo rbtree.lo regional.lo rtt.lo dnstree.lo lookup3.lo lruhash.lo slabhash.lo tcp_conn_limit.lo timehist.lo tube.lo winsock_event.lo autotrust.lo val_anchor.lo rpz.lo rfc_1982.lo proxy_protocol.lo validator.lo val_kcache.lo val_kentry.lo val_neg.lo val_nsec3.lo val_nsec.lo val_secalgo.lo val_sigcrypt.lo val_utils.lo dns64.lo  authzone.lo         respip.lo timeval_func.lo netevent.lo listen_dnsport.lo outside_network.lo keyraw.lo sbuffer.lo wire2str.lo parse.lo parseutil.lo rrdef.lo str2wire.lo  -rpath /usr/local/lib -lssl -lcrypto -lnghttp2
libtool: compile:  gcc -I. -I/opt/ssl33/include -I/usr/local/include -DSRCDIR=. -g -O2 -c smallapp/unbound-anchor.c  -fPIC -DPIC -o .libs/unbound-anchor.o
libtool: compile:  gcc -I. -I/opt/ssl33/include -I/usr/local/include -DSRCDIR=. -g -O2 -c smallapp/unbound-control.c -o unbound-control.o >/dev/null 2>&1
libtool: link: echo "{ global:" > .libs/libunbound.ver
libtool: link:  cat ./libunbound/ubsyms.def | sed -e "s/\(.*\)/\1;/" >> .libs/libunbound.ver
libtool: link:  echo "local: *; };" >> .libs/libunbound.ver
libtool: link:  gcc -shared  -fPIC -DPIC  .libs/context.o .libs/libunbound.o .libs/libworker.o .libs/ub_event_pluggable.o .libs/dns.o .libs/infra.o .libs/rrset.o .libs/dname.o .libs/msgencode.o .libs/as112.o .libs/msgparse.o .libs/msgreply.o .libs/packed_rrset.o .libs/iterator.o .libs/iter_delegpt.o .libs/iter_donotq.o .libs/iter_fwd.o .libs/iter_hints.o .libs/iter_priv.o .libs/iter_resptype.o .libs/iter_scrub.o .libs/iter_utils.o .libs/localzone.o .libs/mesh.o .libs/modstack.o .libs/view.o .libs/outbound_list.o .libs/alloc.o .libs/config_file.o .libs/configlexer.o .libs/configparser.o .libs/fptr_wlist.o .libs/siphash.o .libs/edns.o .libs/locks.o .libs/log.o .libs/mini_event.o .libs/module.o .libs/net_help.o .libs/random.o .libs/rbtree.o .libs/regional.o .libs/rtt.o .libs/dnstree.o .libs/lookup3.o .libs/lruhash.o .libs/slabhash.o .libs/tcp_conn_limit.o .libs/timehist.o .libs/tube.o .libs/winsock_event.o .libs/autotrust.o .libs/val_anchor.o .libs/rpz.o .libs/rfc_1982.o .libs/proxy_protocol.o .libs/validator.o .libs/val_kcache.o .libs/val_kentry.o .libs/val_neg.o .libs/val_nsec3.o .libs/val_nsec.o .libs/val_secalgo.o .libs/val_sigcrypt.o .libs/val_utils.o .libs/dns64.o .libs/authzone.o .libs/respip.o .libs/timeval_func.o .libs/netevent.o .libs/listen_dnsport.o .libs/outside_network.o .libs/keyraw.o .libs/sbuffer.o .libs/wire2str.o .libs/parse.o .libs/parseutil.o .libs/rrdef.o .libs/str2wire.o   -Wl,-rpath -Wl,/opt/ssl33/lib64 -L/opt/ssl33/lib64 -L/usr/local/lib -lssl -lcrypto /usr/local/lib/libnghttp2.so  -g -O2   -Wl,-soname -Wl,libunbound.so.8 -Wl,-version-script -Wl,.libs/libunbound.ver -o .libs/libunbound.so.8.1.26
/usr/bin/ld: /opt/ssl33/lib64/libcrypto.a(libcrypto-lib-ui_openssl.o): warning: relocation against `stdin@@GLIBC_2.2.5' in read-only section `.text'
/usr/bin/ld: /opt/ssl33/lib64/libcrypto.a(libcrypto-lib-cryptlib.o): relocation R_X86_64_PC32 against symbol `stderr@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
make: *** [Makefile:366: libunbound.la] Error 1
make: *** Waiting for unfinished jobs....
libtool: compile:  gcc -I. -I/opt/ssl33/include -I/usr/local/include -DSRCDIR=. -g -O2 -c smallapp/unbound-anchor.c -o unbound-anchor.o >/dev/null 2>&1
```

OS is: Linux `Ubuntu 24.04`.

This is the unbound `configure` I used:

```configure --disable-flto --sysconfdir=/etc --enable-tfo-client --enable-tfo-server  --with-libnghttp2 --with-ssl=/opt/ssl33```

`OpenSSL` was built and available in `/opt/ssl33` and ths is the version:

```
root@server# /opt/ssl33/bin/openssl version
OpenSSL 3.3.0 9 Apr 2024 (Library: OpenSSL 3.3.0 9 Apr 2024)
```
`nghttp2` version is 1.61.0

More, I have the same errors also if I try to use `OpenSSL` version 3.2.1.

Thanks.

**Comment by kkkgo** [comment]
Although I don't use Ubuntu, it compiles well on Alpine Linux. Here is an example script:
```shell
#!/bin/sh
# install package
apk add build-base flex byacc musl-dev gcc make git python3-dev swig libevent-dev openssl-dev expat-dev bind-tools nghttp2-dev
# clone code
git clone https://github.com/NLnetLabs/unbound.git --depth 1 /unbound
cd /unbound || exit
# configure
 ./configure --disable-flto --sysconfdir=/etc --enable-tfo-client --enable-tfo-server  --with-libnghttp2 --with-ssl
# make
make && make install

```
```shell
root@# unbound -V
Version 1.20.1

Configure line: --disable-flto --sysconfdir=/etc --enable-tfo-client --enable-tfo-server --with-libnghttp2 --with-ssl
Linked libs: mini-event internal (it uses select), OpenSSL 3.3.0 9 Apr 2024
Linked modules: dns64 respip validator iterator
TCP Fastopen feature available

BSD licensed, see LICENSE in source package for details.
Report bugs to unbound-bugs@nlnetlabs.nl or https://github.com/NLnetLabs/unbound/issues
```

**Comment by iz8mbw** [comment]
Ok, I found the problem.
The issue was, for some reasons, OpenSSL didn't handle well the "rpath" during `configure` (since I built OpenSSL in a custom path to have Shared Libraries in non-default locations).

I built OpenSSL like this: `config -Wl,-rpath=/opt/ssl33/lib -Wl,--enable-new-dtags --prefix=/opt/ssl33 --openssldir=/opt/ssl33`.
For some reasons the relative path `-Wl,-rpath=/opt/ssl33/lib` was not correclty handled and this was causing isssues on both nghttp2 and unbound.

I rebuild OpenSSL with `config '-Wl,-rpath,$(LIBRPATH)' -Wl,--enable-new-dtags --prefix=/opt/ssl33 --openssldir=/opt/ssl33` and this fixed the OpenSSL builiding issue.
After I re-built nghttp2 and unbound and all was OK.

More info here: https://github.com/openssl/openssl/discussions/24365
Some nores here: https://github.com/openssl/openssl/blob/master/NOTES-UNIX.md#shared-libraries-and-installation-in-non-default-locations
